docs/specs/src/_hwp_D_rate_limiting_spec.md

```diff
@@ -0,0 +1,486 @@
+
+Even though these elements may be arbitrarily expensive to process, the overall system relies on them to function. If a Class A element was rejected due to rate limiting the system could be put into an unrecoverable state as it would be unclear how/when to back off and retry these processes.
+
+Importantly these elements already cannot be spammed by their nature, they do not require additional rate limiting. For example, DNA can only exist once by definition.
+
+Notably the deletion of CapGrants and CapClaims is never rate limited:
+
+- Deletion is always 1:1 with a Create/Update so rate limited creation implies rate limited deletion
+- Rejecting the revocation of access would imply forcing the agent to accept incoming connections they explicitly denied, this is unacceptable for a secure system
+
+However, the deletion entries are private so the network has no way of knowing _which_ deletes line up with what create/updated.
+
+The network does know that in aggregate the following invariant must hold true:
+
+`(deletes + updates) <= (creates + updates)`
+
+Because:
+
+`update = create + delete`
+
+`deletes <= creates`
+
+At the limit where everything is deleted:
+
+`deletes + update-deletes = creates + update-creates`
+
+And each `update-delete` and `update-create` is a single entry therefore:
+
+`(deletes + updates) <= (creates + updates)`
+
+The agent activity neighbourhood can enforce this invariant for each agent.
+
+Note: This is only true of the special case of cap entries as it is invalid for a cap Update/Delete to reference an entry on a different chain so we have a bounded linear history. In the general case entries _can_ reference other chains so the above invariant does not hold. See below for how other entries are handled.
+
+##### Note re: Agent
+
+Many Agents can be created unlike the other Class A entries.
+
+The rate limiting here is implied because all Agent keys are referencing entries in the `deepkey` happ.
+
+Deepkey itself uses rate limiting as specced in this document to limit the rate of Agent entries, so implicitly this is Class A _and_ has rate limit logic.
+
+
+#### B: System priority: Security
+
+Class B elements are access control so considered critical to the system and have high priority but are abusable.
+
+Access control changes are ideally never rate limited:
+
+- False positives leading to inappropriate rate limiting could be very harmful to the user
+- These elements are private entries so the overhead to the network is only the headers (e.g. assigned grants cannot flood the network with garbage keys because this lives in the private entry)
+- Still, there must be _some_ limit, the network cannot accept infinite headers
+
+The _rate_ of cap grants is limited but weight and size (common to all entry headers) are ignored.
+
+As this is entirely a question of network health, a happ should never be intentially (un)limiting access controls. The validation can be done by the system and it is not configurable.
```

Comment:
Not sure if this is correct:

```suggestion
As this is entirely a question of network health, a happ should never be intentionally (un)limiting access controls. The validation can be done by the system and it is not configurable.
```

Reply:
dang, you're really going through this carefully!